### PR TITLE
Fix SocketsHttpHandler system proxy usage logic

### LINKED ISSUE #33866: SocketsHttpHandler Doesn't Respect Manual Proxy Bypass
## Issue

.Net Core versions tested: `2.1.2`, `2.1.6`, `2.2.0`
OS tested: Windows Server 2016 x64

If your Windows proxy settings are set to:

image
image

then `SocketsHttpHandler` doesn't respect proxy bypass preferences.

## Reproduction

- In Internet Options:
  - Tick "Automatically detect settings"
  - Populate the proxy server details at the bottom
  - Click 'Advanced' and add some manual proxy exclusions
- Using a `HttpClient`, request a URL which should match the manual exclusion list.

**Expected result**: The request does not go via the proxy as it matches the bypass list
**Actual Result**: The bypass list is ignored and the request goes via the proxy

## Cause

With the configuration above, `WinInetProxyHelper` has the following properties:

```
WinInetHttpProxyHelper:
        AutoConfigUrl:
        AutoDetect: True
        AutoSettingsUsed: True
        ManualSettingsOnly: False
        Proxy: myproxy.local:8080
        ProxyBypass: localhost;*.local
```

When `WinInetProxyHelper.GetProxyForUrl()` is called, [it fails during auto-detection](https://github.com/dotnet/corefx/blob/0bfc4f3cf62f6fc29162ad5b1b89acc14124a375/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs#L164) since there's no PAC file, and then [falls back to the manual proxy settings](https://github.com/dotnet/corefx/blob/0bfc4f3cf62f6fc29162ad5b1b89acc14124a375/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs#L179) (in this case `myproxy.local:8080`). The manual settings are then [parsed and returned by `HttpSystemProxy.GetProxy()`](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpSystemProxy.cs#L338)

 However, `HttpSystemProxy.IsBypassed` [contains an assumption that `HttpSystemProxy.GetProxy()` will have returned `null` when `AutoSettingsUsed` is `true`](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpSystemProxy.cs#L367), but that assumption is incorrect because the manual fallback settings were returned instead of `null`. This means the URL goes via the proxy even if it matches the bypass list.

## Workaround

You can either:

- Untick "Automatically detect settings" to cause `AutoSettingsUsed` to be `false`, then `HttpSystemProxy.GetProxy()` will actually check the bypass list and return `null` if a bypass matches, or
- Set the environment variable `DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLER` to `0` so that [`SocketsHttpHandler` isn't used](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Net.Http/src/System/Net/Http/HttpClientHandler.Windows.cs#L37) (which shows that this is a regression from the older `WinHttpHandler` implementation).

The longer term solution should probably be that if auto-detection is enabled but fails, set `AutoSettingsUsed` to false so that the normal manual behaviour is used.

### PR DESCRIPTION
**Pull request by davidsh**
SocketsHttpHandler was not using the proxy bypass list specified in IE settings in cases where
there is a combination of both 'AutoDetect' and 'Manual' settings in the IE settings dialog.

SocketsHttpHandler uses WinHttpHandler's WinInetProxyHelper class as part of the HttpSystemProxy
class. But it was consuming it in an incorrect way. WinInetProxyHelper was originally written to
be used only with the rest of the WinHTTP stack for WinHttpHandler. When WinHTTP GetProxyForUrl()
was returning a ProxyBypass string, HttpSystemProxy was ignoring it. It was assuming that the
string for Proxy was correct. But in cases where ProxyBypass is returned, the Proxy string is only
used if the destination uri doesn't match any of the strings in the ProxyBypass list. That logic
would normally be handled automatically by WinHttpHandler. But HttpSystemProxy was simply
discarding the ProxyBypass string returned by WinHTTP GetProxyForUrl().

In order to address this fix, I added new tests for HttpSystemProxy. I utilized the existing mock
layers for registry and WinHTTP interop that are contained in the WinHttpHandler Unit Tests. It might
seem weird to have tests for the internal HttpSystemProxy class located in the WinHttpHandler folder.
But, we already pull in the source code for WinInetProxyHelper from WinHttpHandler into the
SocketsHttpHandler compilation. Using these Unit Tests with the mocking layers allows us to test a
broad range of scenarios (registry settings and network failures) that are normally difficult to test.
I have a plan, though, to refactor the system proxy code and tests as part of implementing #36553.
That will result in the system proxy code and tests ending up in a more logical place.

Fixes #33866

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpSystemProxy.cs [hunk 1]
```diff
@@ -275,7 +275,48 @@ private static int GetProxySubstringLength(string proxyString, int idx)
...
+            if (_proxyHelper.AutoSettingsUsed && !_proxyHelper.RecentAutoDetectionFailure)
+            {
+                var proxyInfo = new Interop.WinHttp.WINHTTP_PROXY_INFO();
+                try
+                {
+                    if (_proxyHelper.GetProxyForUrl(_sessionHandle, uri, out proxyInfo))
+                    {
+                        // If WinHTTP just specified a Proxy with no ProxyBypass list, then
+                        // we can return the Proxy uri directly.
+                        if (proxyInfo.ProxyBypass == IntPtr.Zero)
```
Comment (wfurt) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpSystemProxy.cs]: So the bug was assumption that in automatic processing, WinHttp would check bypass list but it did not. 
And here we expect that the proxyInfo.ProxyBypass would give back same string as we read from registry on first use.

**Comment by davidsh** [comment]
I discovered an unfortunate .NET Framework era behavior of IWebProxy.GetProxy(destinationUri).

Instead of returning null where there isn't a proxy for the destinationUri, it actually returns the destinationUri! 

https://github.com/Microsoft/referencesource/blob/aaca53b025f41ab638466b1efe569df314f689ea/System/net/System/Net/webproxy.cs#L597-L599

```c#
if (AreAllBypassed(proxies, true)) {
    // this is the broken behaviour of IWebProxy. Returning the same destination means bypass
    proxyUri = destination;
```

The .NET Core implementation behaves this way also:
https://github.com/dotnet/corefx/blob/master/src/System.Net.WebProxy/src/System/Net/WebProxy.cs#L87-L95

The documentation for IWebProxy.GetProxy() makes no mention at all of what it does when there is no proxy for a specified Uri. We will have to update that documentation to make things clearer.

However, in our SocketsHttpHandler, we currently handle a `null` return value from IWebProxy.GetProxy(). But we don't handle getting the destinationUri back. But the previous code always called IsByPassed() before calling GetProxy(). So, that logic would work around it. My current PR removed the call to IsByPassed() as an optimization. But now that isn't possible anymore due to the defacto behavior of IWebProxy.

I will revise this PR to take all these items into consideration.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 2]
```diff
@@ -57,29 +57,20 @@ public WinInetProxyHelper()
...
 
-        public bool ManualSettingsOnly
-        {
-            get
-            {
-                return !AutoDetect && string.IsNullOrEmpty(AutoConfigUrl) && !string.IsNullOrEmpty(Proxy);
-            }
-        }
+        public bool ManualSettingsOnly =>
+            !AutoDetect && string.IsNullOrEmpty(AutoConfigUrl) && !string.IsNullOrEmpty(Proxy);
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Nit: maybe:
```C#
public bool ManualSettingsOnly => !AutoSettingsUsed && ManualSettingsUsed;
```
?

Comment (davidsh) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: I didn't have ManualSettingsUsed before this PR.  But now that I do, I can use your suggestion.  Thanks!

**Pull request merged**